### PR TITLE
Add CRD namespace migration draft proposal

### LINKED ISSUE #6471: Migrate CRDs away from *.k8s.io
**User Story**
Kubernetes began to protect the *.k8s.io and *.kubernetes.io namespaces for API groups, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2337-k8s.io-group-protection -- these restrictions have been added in Kubernetes 1.16 (see https://github.com/kubernetes/kubernetes/pull/79992).

The KEP clearly discourages misusing the API groups and says 

> If the API is unapproved, you may set the annotation to a string starting with "unapproved". For instance, "unapproved, temporarily squatting or "unapproved, experimental-only". This is discouraged.

Anticipating that at some point Kubernetes will actually forbid these hacks, we should sooner than later migrate our CRDs away. The KKP Operator already uses `operator.kubermatic.io`, so we should probably converge on `kubermatic.k8c.io`.

**Acceptance criteria**
- [ ] KKP does not use any of the restricted API group domains.
- [ ] Existing KKP setups can be migrated to the new CRDs.

### PR DESCRIPTION
**Pull request by moelsayed**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partial Fix for #6471

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: docs/proposals/crd-namespace-migration.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# CRD Group Migration to K8c.io
+
+**Author**: Mohamed Elsayed (@moelsayed)
+
+**Status**: Draft proposal;
+
+## Goals
+Kubernets API groups are organized by namespaces. Most of KPP CRDs are currently using the `k8s.io` namespace. 
+
+The `k8s.io` namespace is owned/managed by the Kubernets community and is currently protected by an API review process. Ath this point, the Kube-API implements checks to validate if a specific CRD is allowed to use this namespace or not. Unfortunately, we are not allowed.
```
Comment (xrstf) [inline docs/proposals/crd-namespace-migration.md]: `At this point`

File: docs/proposals/crd-namespace-migration.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+**Author**: Mohamed Elsayed (@moelsayed)
+
+**Status**: Draft proposal;
+
+## Goals
+Kubernets API groups are organized by namespaces. Most of KPP CRDs are currently using the `k8s.io` namespace. 
+
+The `k8s.io` namespace is owned/managed by the Kubernets community and is currently protected by an API review process. Ath this point, the Kube-API implements checks to validate if a specific CRD is allowed to use this namespace or not. Unfortunately, we are not allowed. 
+
+This proposal plans to migrate all KPP CRDs from the `k8s.io` namespace into the `kubermatic.io` namespace.
```
Comment (xrstf) [inline docs/proposals/crd-namespace-migration.md]: Let's discuss this. What about `k8c.io`? Short, but maybe confusing (w.r.t. `k8s.io`)?

File: docs/proposals/crd-namespace-migration.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+## Non-Goals
+
+TBA
+
+## Implementation
+
+Implementing this is relatively complex using controller-runtime. It makes more sense to implement this logic in the simplest possible way and try to keep it contained since it should only run once per existing deployment. Newly deployed setups should not have any problems since they will start using the new CRDs from the get go.
+
+The plan is to use client-go dynamic clients to implement the migration code. I deally it will be included the `kubermatic-installer` and It should apply the following steps during upgrades for existing deployments:
```
Comment (xrstf) [inline docs/proposals/crd-namespace-migration.md]: `Ideally`

Comment (xrstf) [reply docs/proposals/crd-namespace-migration.md]: lowercase `it`

File: docs/proposals/crd-namespace-migration.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+- Deploy the new CRDs with the updated group namespace. At this point the KPP deployment will contain _both_ CRDs.
+- List all existing resources with old group namespace
+- Copy over the existing resources with the new group namespace. At this point, the old KPP controllers are still running and will not reconcile the newly created resources.
```
Comment (xrstf) [inline docs/proposals/crd-namespace-migration.md:26]: Before we do this, we need to shut KKP down, so no changes can occur to the old resources anymore.

File: docs/proposals/crd-namespace-migration.md (hunk 2 above)
Comment (moelsayed) [reply docs/proposals/crd-namespace-migration.md]: I actually wrote `k8c.io` here first, then changed it to `kubermatic.io` when I noticed your note about having that already for the operator. My +1 is actually for `k8c.io`

File: docs/proposals/crd-namespace-migration.md (hunk 4 above)
Comment (moelsayed) [reply docs/proposals/crd-namespace-migration.md:26]: Shut the whole thing down or just the API would be enough?

Comment (xrstf) [reply docs/proposals/crd-namespace-migration.md:26]: Everything I would say. Just to be on the safe side.

File: docs/proposals/crd-namespace-migration.md (hunk 2 above)
Comment (xrstf) [reply docs/proposals/crd-namespace-migration.md]: Apparently we had already discussed this previously and settled on `k8c.io`, according to my team.

**Comment by xrstf** [comment]
One major thing we should talk about as well: Currently, the KKP CRDs are a little bit chaotic. Some are global, some are namespaced. We should clean it up and make sure it's easy/easier for admins to know where to look for resources.

**Comment by moelsayed** [comment]
> One major thing we should talk about as well: Currently, the KKP CRDs are a little bit chaotic. Some are global, some are namespaced. We should clean it up and make sure it's easy/easier for admins to know where to look for resources.

Perfect for the `Non-goals` section :joy:. I think this would be out of scope for this proposal. I understand it's kind of related, but it's a much bigger change and it's not as urgent as the CRD namespace issue.

**Pull request merged**